pkg/lib/infra/sms/sendcloud/models/send.go

```diff
@@ -0,0 +1,109 @@
+package models
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strings"
+)
+
+type SendRequest struct {
+	msgType       string
+	phone         []string
+	sendRequestId string
+	smsUser       string
+	templateId    string
+	vars          map[string]interface{}
+}
+
+func NewSendRequest(
+	msgType string,
+	phone []string,
+	smsUser string,
+	templateId string,
+	vars map[string]interface{},
+) SendRequest {
+	s := SendRequest{
+		msgType:    msgType,
+		phone:      phone,
+		smsUser:    smsUser,
+		templateId: templateId,
+		vars:       vars,
+	}
+	presign := s.Presign()
+	h := md5.Sum([]byte(presign))
+	sendRequestId := hex.EncodeToString(h[:])
+	return SendRequest{
+		msgType:       msgType,
+		phone:         phone,
+		sendRequestId: sendRequestId,
+		smsUser:       smsUser,
+		templateId:    templateId,
+		vars:          vars,
+	}
+}
+
+func (r *SendRequest) Presign() string {
+	vars, _ := json.Marshal(r.vars)
+	return strings.Join([]string{
+		fmt.Sprintf("msgType=%v", r.msgType),
+		fmt.Sprintf("phone=%v", strings.Join(r.phone, ",")),
+		fmt.Sprintf("sendRequestId=%v", r.sendRequestId),
+		fmt.Sprintf("smsUser=%v", r.smsUser),
+		fmt.Sprintf("templateId=%v", r.templateId),
+		fmt.Sprintf("vars=%v", string(vars)),
+	}, "&")
```

Comment:
Not sure if the signature requires a specific order of the arguments.
Not sure if the signature requires you to use a x-www-form-urlencoded, if yes, then the value should be encoded with url.QueryEscape

Reply:
The arguments need to be alphabetical order

https://www.sendcloud.net/doc/sms/
> 1. 将实际调用API的参数以字母升序(A-Z)排列, 不包括 smsKey 和 signature 字段本身

The values in signature not need to be url encoded.

https://www.sendcloud.net/doc/sms/
> 1. 生成签名时, 参数不要使用 'urlencode'. 在调用 api 时, 才需要对参数做 'urlencode'

Reply:
Then you should include these in the comments. For a reader like me, I have no clue what you write is intended or not.

Reply:
Comments added via 5e2df93